README.md
<div align="center">
  <a href="https://github.com/banyancomputer/fake-file" target="_blank">
    <img src="https://raw.githubusercontent.com/banyancomputer/fake-file/main/assets/a_logo.png" alt="fake-file Logo" width="100"></img>
  </a>

  <h1 align="center">fake-file</h1>
</div>

<div align="center"><sub>:warning: Work in progress :warning:</sub></div>

##

## Outline

- [Installation](#installation)
- [Testing the Project](#testing-the-project)
- [Benchmarking the Project](#benchmarking-the-project)
- [Getting Help](#getting-help)
- [External Resources](#external-resources)
- [License](#license)

## Installation

### Using `cargo`

```console
cargo install fake-file
```

## Usage
```console
# Get the help menu
fake-file --help
# Generate a 4 level deep directory structre. Each directory should have 4 files.
# Each end file should add up to 1024 bytes (1 KB) in size.
fake-file -d 4 -w 4 -s 1024 -o .
```

## Testing the Project

- Run tests

  ```console
  cargo test
  ```

## Benchmarking the Project

- Run benchmarks - this sees how long it takes to generate a 1GB file structure

  ```console
  cargo bench 
  ```

### Formatting

For formatting Rust in particular, please use `cargo fmt` as it uses
specific nightly features we recommend. **Make sure you have nightly
installed**.

## Getting Help

For usage questions, usecases, or issues please open an issue in our repository.

We would be happy to try to answer your question or try opening a new issue on Github.

## External Resources

These are references to specifications, talks and presentations, etc.

## License

This project is licensed under the [MIT License](./LICENSE),
or [http://opensource.org/licenses/MIT][mit].


[cargo-expand]: https://github.com/dtolnay/cargo-expand
[cargo-udeps]: https://github.com/est31/cargo-udeps
[cargo-watch]: https://github.com/watchexec/cargo-watch
[commit-spec]: https://www.conventionalcommits.org/en/v1.0.0/#specification
[commit-spec-site]: https://www.conventionalcommits.org/
[criterion]: https://github.com/bheisler/criterion.rs
[irust]: https://github.com/sigmaSd/IRust
[mit]: http://opensource.org/licenses/MIT
[pre-commit]: https://pre-commit.com/
[proptest]: https://github.com/proptest-rs/proptest
[strategies]: https://docs.rs/proptest/latest/proptest/strategy/trait.Strategy.html


CHANGELOG.md
# Changelog

## [0.1.1](https://github.com/banyancomputer/fake-file/compare/fake-file-v0.1.0...fake-file-v0.1.1) (2023-04-10)


### Features

* better output ([a886869](https://github.com/banyancomputer/fake-file/commit/a8868698641b36f5d986b10cc759254135b38390))
* cli ([efb4a4f](https://github.com/banyancomputer/fake-file/commit/efb4a4f436e8dbe8f52c250e37cfae0828e75cb1))
* fake-file ported to its own crate as lib ([7978f41](https://github.com/banyancomputer/fake-file/commit/7978f41ea882fd98c4c82b0367373792272589cd))
* useful topology customization ([7038bfb](https://github.com/banyancomputer/fake-file/commit/7038bfb8fba6f77bb12c374d15fd7caff220dd0c))
* useful topology customization ([d105d39](https://github.com/banyancomputer/fake-file/commit/d105d3924b38a2e444f028850dabc7a5e30561de))
* verbose option ([8050f4f](https://github.com/banyancomputer/fake-file/commit/8050f4ff2bb502c36f3f4ad9cef32b0f48c64cc4))
* verbose option ([ee5d80d](https://github.com/banyancomputer/fake-file/commit/ee5d80d1d8b6a044a524e102d83c25b4a933f0bc))


### Bug Fixes

* directory creates a directory ([4cbaa61](https://github.com/banyancomputer/fake-file/commit/4cbaa618b8849af4f6b13a37eb5baebda08a096f))
* mutable references are stupid ([0355bf0](https://github.com/banyancomputer/fake-file/commit/0355bf05cdb4adb373c09174e171127dc02be2da))

## Changelog


tests/balanced.rs
#[cfg(test)]
mod test {
    use fake_file::{Strategy, Structure};
    use fs_extra::dir::get_size;
    use std::{fs, path::Path};

    const TEST_SCRATCH_SPACE: &str = "test";
    const TEST_SIZE: usize = 1024 * 1024;
    const TEST_WIDTH: usize = 2;
    const TEST_DEPTH: usize = 2;

    /// Create a balanced file structure, 1 KB in size
    #[test]
    fn test_balanced() {
        let test_scratch_space = Path::new(TEST_SCRATCH_SPACE);
        // Remove the scratch space and recreate it
        fs::remove_dir_all(&test_scratch_space).unwrap_or(());
        fs::create_dir_all(&test_scratch_space).unwrap();
        // Create a file structure
        let structure = Structure::new(TEST_WIDTH, TEST_DEPTH, TEST_SIZE, Strategy::Balanced);
        // Push another path onto the scratch space
        let test_scratch_space = test_scratch_space.join(structure.to_path_string());
        // Generate the file structure
        structure
            .generate(&test_scratch_space)
            .map_err(|e| {
                println!("Error Generating FS: {}", e);
            })
            .unwrap();
        // Check that the file structure was created
        assert!(test_scratch_space.exists());
        // Check the the file structure is around the right size
        let structure_size = get_size(&test_scratch_space).unwrap();

        assert_eq!(structure_size, TEST_SIZE as u64);
    }
}


tests/structure.rs
#[cfg(test)]
mod test {
    use fake_file::{Strategy, Structure};
    use fs_extra::dir::get_size;
    use std::{fs, path::Path};

    const TEST_SCRATCH_SPACE: &str = "test";
    const TEST_SIZE: usize = 1024 * 1024;
    const TEST_WIDTH: usize = 2;
    const TEST_DEPTH: usize = 2;

    /// Create a balanced file structure, 1 KB in size
    #[test]
    fn test_simple() {
        let test_scratch_space = Path::new(TEST_SCRATCH_SPACE);
        // Remove the scratch space and recreate it
        fs::remove_dir_all(&test_scratch_space).unwrap_or(());
        fs::create_dir_all(&test_scratch_space).unwrap();
        // Create a file structure
        let structure = Structure::new(TEST_WIDTH, TEST_DEPTH, TEST_SIZE, Strategy::Simple);
        // Push another path onto the scratch space
        let test_scratch_space = test_scratch_space.join(structure.to_path_string());
        // Generate the file structure
        structure
            .generate(&test_scratch_space)
            .map_err(|e| {
                println!("Error Generating FS: {}", e);
            })
            .unwrap();
        // Check that the file structure was created
        assert!(test_scratch_space.exists());
        // Check the the file structure is around the right size
        let structure_size = get_size(&test_scratch_space).unwrap();

        assert_eq!(structure_size, TEST_SIZE as u64);
    }
}


src/lib.rs
#![cfg_attr(docsrs, feature(doc_cfg))]
#![warn(missing_debug_implementations, rust_2018_idioms)]
#![deny(unreachable_pub, private_in_public)]

use crate::utils::create_random_file;
use anyhow::{Error, Result};
use serde::{Deserialize, Serialize};
use std::{fs, path::Path};

// fake-file is a library for generating random file structures with certain properties.

/// Crate utility functions
pub mod utils;

/// Enum for describing how to generate a file structure
/// This is used for generating random data for testing
#[derive(Serialize, Deserialize, Clone, Debug, strum::Display)]
pub enum Strategy {
    /// Generate exactly the file structure requested
    Simple,
    /// Generate a wide version of the file structure requested -- half as deep, twice as wide
    Wide,
    /// Generate a deep version of the file structure requested -- twice as deep, half as wide
    Deep,
    /// Generate a very shallow version of the file structure requested -- `wide` files in a single directory
    Directory,
    /// Generate one big file with the requested size at the specified path
    File,
}

/// Everything is a file in Unix :) including directories
/// Struct for representing a file structure, regardless of depth (i.e. a file or a directory)
/// We use this for generating random data for testing
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Structure {
    /// How many files should be in the file (if it has depth > 0)
    pub width: usize,
    /// How deep the directory structure should be. 0 means this is a file
    pub depth: usize,
    // TODO: This should be Kb in a future version
    /// How much data should be in the file, in bytes
    pub target_size: usize,
}

impl std::str::FromStr for Strategy {
    type Err = String;
    /// Convert a string to a Strategy
    /// # Arguments
    /// s: The string to convert
    /// # Returns
    /// `Result<Strategy>`
    /// # Panics
    /// Panics if the string is not a valid strategy
    /// # Examples
    /// ```no_run
    /// use fake_file::Strategy;
    /// use std::str::FromStr;
    /// let strategy = Strategy::from_str("Simple").unwrap();
    /// ```
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "Simple" => Ok(Strategy::Simple),
            "Wide" => Ok(Strategy::Wide),
            "Deep" => Ok(Strategy::Deep),
            "Directory" => Ok(Strategy::Directory),
            "File" => Ok(Strategy::File),
            _ => Err(format!("Invalid strategy: {}", s)),
        }
    }
}

impl Structure {
    /// Create a new FileStructure
    /// # Arguments
    /// width: Desired width of the file structure
    /// depth: Desired depth of the file structure
    /// target_size: Desired size of the file structure
    /// strategy: The strategy to use for generating the file structure
    /// # Returns
    /// FileStructure
    /// # Examples
    /// ```no_run
    /// use fake_file::{Structure, Strategy};
    /// let s = Structure::new(
    ///    4,                               // width
    ///   4,                               // depth
    ///   1024,                            // target_size
    ///   Strategy::Simple,                // strategy
    /// );
    pub fn new(width: usize, depth: usize, target_size: usize, strategy: Strategy) -> Self {
        match strategy {
            Strategy::Wide => Structure {
                width: width * 2,
                depth: depth / 2,
                target_size,
            },
            Strategy::Deep => Structure {
                width: width / 2,
                depth: depth * 2,
                target_size,
            },
            Strategy::Directory => Structure {
                width: width,
                depth: 1,
                target_size,
            },
            Strategy::File => Structure {
                width: 1,
                depth: 1,
                target_size,
            },
            _ => Structure {
                width,
                depth,
                target_size,
            },
        }
    }

    /// Convert the FileStructure to a string that can be used as a filename
    /// # Example
    /// ```no_run
    /// use fake_file::{Structure, Strategy};
    /// let s = Structure::new(
    ///    4,                               // width
    ///   4,                               // depth
    ///  1024 * 1024,
    /// Strategy::Simple                   // target size in bytes (1Mb)
    /// );
    /// assert_eq!(s.to_path_string(), "w4_d4_s1048576");
    /// ```
    pub fn to_path_string(&self) -> String {
        format!("w{}_d{}_s{}", self.width, self.depth, self.target_size)
    }

    /// Generate a FileStructure with the given path. Does not check if the path can hold
    /// the file structure. Use with caution!
    /// # Arguments
    /// path: The path to generate the file structure at
    /// # Panics
    /// Panics if it can't create a directory at the given path (i.e. the path parent doesn't exist)
    /// Panics if the given path already exists
    /// Errors if the file structure cannot be generated for some reason
    /// # Example
    /// ```no_run
    /// use fake_file::{Structure, Strategy};
    /// use std::path::Path;
    /// let s = Structure::new(
    ///   4,                               // width
    ///  4,                               // depth
    /// 1024 * 1024,                      // target size in bytes (1Mb)
    /// Strategy::Simple,
    /// );
    /// let path = Path::new("/tmp/fake_file");
    /// s.generate(path).unwrap();
    /// ```
    pub fn generate(&self, path: &Path) -> Result<(), Error> {
        // Panic if the path already exists. We don't want to overwrite anything!
        assert!(!path.exists());
        // If this is 0, we're creating a file
        if self.depth == 0 {
            // let file_path = path;
            // // Create a file with the target size

            create_random_file(path, self.target_size);
            return Ok(()); // We're done here
        }
        fs::create_dir(path).unwrap();
        for i in 0..self.width {
            // Read a fixed amount of data from target size
            let target_size = self.target_size / self.width;
            // Push the new path onto the path
            // let mut new_path = path.clone();
            let new_path = path.join(i.to_string());
            // Generate a new FileStructure with the new path
            Structure::new(self.width, self.depth - 1, target_size, Strategy::Simple)
                .generate(&new_path)
                .unwrap();
        }
        Ok(())
    }
}


src/main.rs
//! fake-file

use clap::Parser;
use fake_file::Structure;
use std::path::PathBuf;

mod cli;

/// Main entry point.
fn main() {
    // Parse command line arguments. see args.rs
    let cli = cli::Args::parse();
    let s = Structure::new(cli.width, cli.depth, cli.size, cli.strategy);

    // If the output path is not specified, use the to_path_string() method
    // to get the current working directory.

    let mut output = PathBuf::from(cli.output);
    output.push(s.to_path_string());

    // If the output path exists
    if output.exists() {
        // If the force flag is not set, exit with an error.
        if !cli.force {
            println!("Output path already exists. Use --force to overwrite.");
            println!("Path: {:?}", output);
            std::process::exit(0);
        }
        if cli.verbose {
            println!("WARN: Output path already exists. Overwriting.");
            println!("Path: {:?}", output);
        }
        // Delete the output path
        std::fs::remove_dir_all(&output).unwrap();
    }

    // If verbose output is enabled, print the structure
    if cli.verbose {
        println!("Generating File Structure: {:?}", s);
        println!("Output Path: {:?}", output);
    }

    let now = std::time::Instant::now();

    s.generate(&output)
        .map_err(|e| {
            println!("Error Generating File Structure: {}", e);
        })
        .unwrap();

    let elapsed = now.elapsed().as_secs_f64();
    if cli.verbose {
        println!("Completed in: {:?} s", elapsed);
    }
}


src/utils.rs
use anyhow::{anyhow, Result};
use rand::Rng;
use std::{
    cmp, fs,
    io::{BufWriter, Write},
    path::Path,
};

// TODO: This should be Kb in a future version
/// Creates a random file at the given path with the given size
/// # Arguments
/// * `path` - The path to create the file at
/// * `size` - The size of the file to create in bytes
/// # Panics
/// Panics if the file cannot be created
/// # Examples
/// ```no_run
/// use fake_file::utils::create_random_file;
/// let path = std::path::Path::new("test.txt");
/// create_random_file(&path, 1024);
/// ```
pub fn create_random_file(path: &Path, size: usize) {
    let file = fs::File::create(path).unwrap();
    let mut rng = rand::thread_rng();
    let mut writer = BufWriter::new(file);

    let mut buffer = [0; 1024];
    let mut remaining_size = size;

    while remaining_size > 0 {
        let to_write = cmp::min(remaining_size, buffer.len());
        let buffer = &mut buffer[..to_write];
        rng.fill(buffer);
        writer.write_all(buffer).unwrap();

        remaining_size -= to_write;
    }
}

/// Ensures that the given path exists and is a directory
/// # Arguments
/// path: The path to check
/// # Returns
/// Creates the directory if it doesn't exist, and is a directory
/// Result<()>
/// # Panics
/// Panics if the path exists but is not a directory
/// # Examples
/// ```no_run
/// use fake_file::utils::ensure_path_exists_and_is_dir;
/// let path = std::path::Path::new("test");
/// ensure_path_exists_and_is_dir(&path).unwrap();
/// ```
pub fn ensure_path_exists_and_is_dir(path: &Path) -> Result<()> {
    if !path.exists() {
        // create path if it doesn't exist
        fs::create_dir_all(path)?;
    }
    if !path.is_dir() {
        return Err(anyhow!("Path is not a directory: {}", path.display()));
    }
    Ok(())
}

/// Ensures that the given path exists and is a directory and is empty
/// # Arguments
/// path: The path to check
/// # Returns
/// Creates the directory if it doesn't exist. Makes the directory empty if it is not empty.
/// Result<()>
/// # Panics
/// Panics if the path is not an existing directory.
/// Panics if the path is not empty and force is false.
/// # Examples
/// ```no_run
/// use fake_file::utils::ensure_path_exists_and_is_empty_dir;
/// let path = std::path::Path::new("test");
/// ensure_path_exists_and_is_empty_dir(&path, false).unwrap();
/// ```
pub fn ensure_path_exists_and_is_empty_dir(path: &Path, force: bool) -> Result<()> {
    // Check the path exists and is a directory
    ensure_path_exists_and_is_dir(path)?;
    // Check the path is empty
    if path.read_dir().unwrap().count() > 0 {
        // If force is true, make the path empty
        if force {
            fs::remove_dir_all(path)?;
            fs::create_dir_all(path)?;
        } else {
            return Err(anyhow!("Path is not empty: {}", path.display()));
        }
    }
    Ok(())
}


benches/structure.rs
use criterion::{black_box, criterion_group, criterion_main, BatchSize, Criterion};
use lazy_static::lazy_static;
use std::{env, fs, path::Path, time};

use fake_file::{
    utils::{ensure_path_exists_and_is_dir, ensure_path_exists_and_is_empty_dir},
    Strategy, Structure,
};

// Configure the Benching Framework from the Environment
lazy_static! {
    // Path we will use to hold our benchmarking data
    static ref BENCH_PATH: String = env::var("BENCH_PATH").unwrap_or_else(|_| "bench".to_string());
}

/// Ready a directory for benchmarking
#[doc(hidden)]
fn prep_generate(path: &Path) {
    // If the path exists, remove it, regardless of whether it is a file or directory
    if path.exists() {
        fs::remove_dir_all(path).unwrap();
    }
}

/// Bench the generation of a file structure
#[doc(hidden)]
fn structure(c: &mut Criterion) {
    // Get the Bench path and make sure it exists
    let bench_scratch_space = Path::new(BENCH_PATH.as_str());
    ensure_path_exists_and_is_dir(&bench_scratch_space).unwrap();
    // Get the input path and make sure it exists and is empty
    let balanced_path = bench_scratch_space.join("balanced");
    ensure_path_exists_and_is_empty_dir(&balanced_path, true).unwrap();
    // Declare a balanced file structure
    let structure = Structure::new(
        4,                  // width
        4,                  // depth
        1024 * 1024 * 1024, // target size in bytes (1Gb)
        Strategy::Simple,   // strategy
    );
    // Get a path to the file structure
    let structure_path = structure.to_path_string();
    // Get the full path to the file structure
    let full_path = balanced_path.join(structure_path);
    // Create the benchmark group
    let mut group = c.benchmark_group("balanced");
    // Add a throughput benchmark
    group.throughput(criterion::Throughput::Bytes(structure.target_size as u64));
    group.bench_function("balanced", |b| {
        b.iter_batched(
            || {
                // Prep the input path
                prep_generate(&full_path);
            },
            |_| {
                // Generate the file structure
                structure.generate(black_box(&full_path)).unwrap();
            },
            BatchSize::PerIteration,
        )
    });
    // Finish the benchmark group
    group.finish()
}

criterion_group! {
    name = benches;
    config = Criterion::default().sample_size(10).measurement_time(time::Duration::from_secs(30));
    targets = structure
}
criterion_main!(benches);


src/utils/fs_utils.rs
use anyhow::{anyhow, Result};
use rand::Rng;
use std::{
    cmp, fs,
    io::{BufWriter, Write},
    path::Path,
};

/// Creates a random file at the given path with the given size
/// # Arguments
/// * `path` - The path to create the file at
/// * `size` - The size of the file to create in bytes
/// # Panics
/// Panics if the file cannot be created
/// # Examples
/// ```no_run
/// use fake_file::utils::fs_utils::create_random_file;
/// let path = std::path::Path::new("test.txt");
/// create_random_file(&path, 1024);
/// ```
pub fn create_random_file(path: &Path, size: usize) {
    let file = fs::File::create(path).unwrap();
    let mut rng = rand::thread_rng();
    let mut writer = BufWriter::new(file);

    let mut buffer = [0; 1024];
    let mut remaining_size = size;

    while remaining_size > 0 {
        let to_write = cmp::min(remaining_size, buffer.len());
        let buffer = &mut buffer[..to_write];
        rng.fill(buffer);
        writer.write_all(buffer).unwrap();

        remaining_size -= to_write;
    }
}

/// Ensures that the given path exists and is a directory
/// # Arguments
/// path: The path to check
/// # Returns
/// Creates the directory if it doesn't exist, and is a directory
/// Result<()>
/// # Panics
/// Panics if the path exists but is not a directory
/// # Examples
/// ```no_run
/// use fake_file::utils::fs_utils::ensure_path_exists_and_is_dir;
/// let path = std::path::Path::new("test");
/// ensure_path_exists_and_is_dir(&path).unwrap();
/// ```
#[doc(hidden)]
pub fn ensure_path_exists_and_is_dir(path: &Path) -> Result<()> {
    if !path.exists() {
        // create path if it doesn't exist
        fs::create_dir_all(path)?;
    }
    if !path.is_dir() {
        return Err(anyhow!("Path is not a directory: {}", path.display()));
    }
    Ok(())
}

/// Ensures that the given path exists and is a directory and is empty
/// # Arguments
/// path: The path to check
/// # Returns
/// Creates the directory if it doesn't exist. Makes the directory empty if it is not empty.
/// Result<()>
/// # Panics
/// Panics if the path is not an existing directory.
/// Panics if the path is not empty and force is false.
/// # Examples
/// ```no_run
/// use fake_file::utils::fs_utils::ensure_path_exists_and_is_empty_dir;
/// let path = std::path::Path::new("test");
/// ensure_path_exists_and_is_empty_dir(&path, false).unwrap();
/// ```
#[doc(hidden)]
pub fn ensure_path_exists_and_is_empty_dir(path: &Path, force: bool) -> Result<()> {
    // Check the path exists and is a directory
    ensure_path_exists_and_is_dir(path)?;
    // Check the path is empty
    if path.read_dir().unwrap().count() > 0 {
        // If force is true, make the path empty
        if force {
            fs::remove_dir_all(path)?;
            fs::create_dir_all(path)?;
        } else {
            return Err(anyhow!("Path is not empty: {}", path.display()));
        }
    }
    Ok(())
}


benches/balanced.rs
use criterion::{black_box, criterion_group, criterion_main, BatchSize, Criterion};
use lazy_static::lazy_static;
use std::{env, fs, path::Path, time};

use fake_file::{
    utils::fs_utils::{ensure_path_exists_and_is_dir, ensure_path_exists_and_is_empty_dir},
    Strategy, Structure,
};

// Configure the Benching Framework from the Environment
lazy_static! {
    // Path we will use to hold our benchmarking data
    static ref BENCH_PATH: String = env::var("BENCH_PATH").unwrap_or_else(|_| "bench".to_string());
}

/// Ready a directory for benchmarking
#[doc(hidden)]
fn prep_generate(path: &Path) {
    // If the path exists, remove it, regardless of whether it is a file or directory
    if path.exists() {
        fs::remove_dir_all(path).unwrap();
    }
}

/// Bench the generation of a file structure
#[doc(hidden)]
fn balanced_structure(c: &mut Criterion) {
    // Get the Bench path and make sure it exists
    let bench_scratch_space = Path::new(BENCH_PATH.as_str());
    ensure_path_exists_and_is_dir(&bench_scratch_space).unwrap();
    // Get the input path and make sure it exists and is empty
    let balanced_path = bench_scratch_space.join("balanced");
    ensure_path_exists_and_is_empty_dir(&balanced_path, true).unwrap();
    // Declare a balanced file structure
    let structure = Structure::new(
        4,                  // width
        4,                  // depth
        1024 * 1024 * 1024, // target size in bytes (1Gb)
        Strategy::Balanced, // Balanced
    );
    // Get a path to the file structure
    let structure_path = structure.to_path_string();
    // Get the full path to the file structure
    let full_path = balanced_path.join(structure_path);
    // Create the benchmark group
    let mut group = c.benchmark_group("balanced");
    // Add a throughput benchmark
    group.throughput(criterion::Throughput::Bytes(structure.target_size as u64));
    group.bench_function("balanced", |b| {
        b.iter_batched(
            || {
                // Prep the input path
                prep_generate(&full_path);
            },
            |_| {
                // Generate the file structure
                structure.generate(black_box(&full_path)).unwrap();
            },
            BatchSize::PerIteration,
        )
    });
    // Finish the benchmark group
    group.finish()
}

criterion_group! {
    name = benches;
    config = Criterion::default().sample_size(10).measurement_time(time::Duration::from_secs(30));
    targets = balanced_structure
}
criterion_main!(benches);
